### DISCUSSION TITLE
[DISREGARD] Decrypt 48 bytes with AES-128 CFB and get 42 bytes?

### DISCUSSION THREAD

**Discussion by darrellroot** [krzyzanowskim/CryptoSwift#971] [General]
EDIT: This discussion can be disregarded.  My application is for SNMPv3 with encryption and per RFC3826 section 3.1.3 SNMPv3-AES uses a special procedure to encrypt/decrypt the last block.
END-EDIT

Here's a test case where I decrypt 48 bytes with AES-128 CFB and get 42 bytes.  They are the correct leading 42 bytes, but missing the last 6 bytes.  This is with Swift 5.6 in Xcode 13.4.1.  Is this a bug?  Or is the problem between keyboard and chair on my end? - Darrell

    func testAesDecode3() throws {
        let encryptedContents: [UInt8] = [0x3e,0x0c,0xb9,0x50,0xf9,0xeb,0x8f,0x85,0xd2,0xfc,0x57,0xf9,0x71,0x2c,0x13,0xa3,0x88,0xab,0x1d,0x38,0x58,0x2a,0x62,0xae,0x38,0xe5,0x85,0xb1,0x9a,0x8b,0x86,0x94,0x4e,0xab,0x90,0x18,0xf8,0xb4,0x79,0x30,0x69,0x4c,0xe1,0xf0,0x3a,0xd4,0xcf,0x99]
        let initializationVector: [UInt8] = [0x00,0x00,0x00,0x03,0x00,0x07,0xae,0xad,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x87]
        let localizedKey: [UInt8] = [0xe5,0x6f,0xb9,0xb8,0x6b,0x1b,0x09,0xf3,0x5f,0xe4,0x34,0x70,0x8b,0x65,0x6d,0x8d]
        let aes = try AES(key: localizedKey, blockMode: CFB(iv: initializationVector))
        let decryptedContents = try aes.decrypt(encryptedContents)
        XCTAssert(encryptedContents.count == 48)
        XCTAssert(initializationVector.count == 16)
        XCTAssert(localizedKey.count == 16)
        let expectedResults: [UInt8] = [0x30,0x2e,0x04,0x0b,0x80,0x00,0x00,0x09,0x03,0x4c,0x71,0x0c,0x19,0xe3,0x0d,0x04,0x00,0xa2,0x1d,0x02,0x04,0x16,0xb6,0x05,0x5a,0x02,0x01,0x00,0x02,0x01,0x00,0x30,0x0f,0x30,0x0d,0x06,0x08,0x2b,0x06,0x01,0x02,0x01,0x01,0x07,0x00,0x02,0x01,0x06]
        XCTAssert(expectedResults.count == 48)
        XCTAssert(encryptedContents.count == decryptedContents.count)
        XCTAssert(decryptedContents == expectedResults)
    }

<img width="1505" alt="Screen Shot 2022-08-04 at 9 00 00 PM" src="https://user-images.githubusercontent.com/34557273/182998937-e3124a18-3ce3-4ac4-aad2-fccb521cdf40.png">

**Comment by krzyzanowskim** [comment]
> My application is for SNMPv3 with encryption and per RFC3826 section 3.1.3 SNMPv3-AES uses a special procedure to encrypt/decrypt the last block.

I'd love to hear more here. In general, I'd expect to decrypt 48 bytes (?) for CFB.

**Reply by darrellroot** [reply]
The problem ended up being padding related.  By adding "padding: .noPadding" to line 461 of my test case above it started working!  I have a full SwiftSnmpWalk test run going now and it's going much better!

        let aes = try AES(key: localizedKey, blockMode: CFB(iv: initializationVector), padding: .noPadding)

Without that your AES initializer was expecting .pkcs7 padding by default.  Since my data was not .pkcs7 padded, bad things happened.

So "problem between keyboard and chair" on my end. ;-)
